modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+    trace_path: TracePath,
+    /// Base denomination of the relayed fungible token.
+    base_denom: Denom,
+}
+
+impl DenomTrace {
+    /// Returns a coin denomination for an ICS20 fungible token in the format
+    /// 'ibc/{Hash(trace_path/base_denom)}'.
+    pub fn hashed(&self) -> HashedDenom {
+        HashedDenom::from(self)
+    }
+
+    /// Removes the specified prefix from the trace path if there is a match, otherwise does nothing.
+    pub fn remove_trace_prefix(&mut self, prefix: &TracePrefix) {
+        self.trace_path.remove_prefix(prefix)
+    }
+
+    /// Adds the specified prefix to the trace path.
+    pub fn add_trace_prefix(&mut self, prefix: TracePrefix) {
+        self.trace_path.add_prefix(prefix)
+    }
+
+    /// Returns `Source::Receiver` if the denomination originally came from the receiving chain and
+    /// `Source::Sender` otherwise.
+    pub fn source_chain(&self, prefix: &TracePrefix) -> Source {
+        if self.trace_path.starts_with(prefix) {
+            Source::Receiver
+        } else {
+            Source::Sender
+        }
+    }
+}
+
+impl FromStr for DenomTrace {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts: Vec<&str> = s.split('/').collect();
+        let last_part = parts.pop().expect("split() returned an empty iterator");
+
+        let (base_denom, trace_path) = {
+            if last_part == s {
+                (Denom::from_str(s)?, TracePath::default())
+            } else {
+                let base_denom = Denom::from_str(last_part)?;
+                let trace_path = TracePath::try_from(parts)?;
+                (base_denom, trace_path)
+            }
+        };
+
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl TryFrom<RawDenomTrace> for DenomTrace {
+    type Error = Error;
+
+    fn try_from(value: RawDenomTrace) -> Result<Self, Self::Error> {
+        let base_denom = Denom::from_str(&value.base_denom)?;
+        let trace_path = TracePath::from_str(&value.path)?;
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl From<DenomTrace> for RawDenomTrace {
+    fn from(value: DenomTrace) -> Self {
+        Self {
+            path: value.trace_path.to_string(),
+            base_denom: value.base_denom.to_string(),
+        }
+    }
+}
+
+impl From<Denom> for DenomTrace {
+    fn from(denom: Denom) -> Self {
+        Self {
+            trace_path: Default::default(),
+            base_denom: denom,
+        }
+    }
+}
+
+impl fmt::Display for DenomTrace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.trace_path.0.is_empty() {
+            write!(f, "{}", self.base_denom)
+        } else {
+            write!(f, "{}/{}", self.trace_path, self.base_denom)
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct HashedDenom(Vec<u8>);
```

Comment:
Should we even have a `HashedDenom`? I find that it adds a lot of complexity to this library (e.g. `IbcCoin` only exists because we have `HashedDenom`), which I believe should only contain the core token transfer logic. IMO, whether or not to hash denoms before storing them is an implementation detail of how chains store denoms, let alone the fact that we're currently forcing chains to deal with hashed denoms. In my view, this fits best in some `utils` library that chains can use if they so choose. 

The one reason I can see that we might want to keep it is so that we can emit the same events as ibc-go, and ibc-go currently emits events containing trace hashes. However, we could also simply hash the denom internally to create the event, and throw the hash away after.